pkg/river/encoding/array.go

```diff
@@ -0,0 +1,87 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/grafana/agent/pkg/river/internal/value"
+)
+
+// ArrayField represents an array node.
+type ArrayField struct {
+	Type         string        `json:"type"`
+	Value        []interface{} `json:"value,omitempty"`
+	valueFields  []*ValueField
+	structFields []*StructField
+	arrayFields  []*ArrayField
+	mapFields    []*MapField
+}
+
+func newArray(val value.Value) (*ArrayField, error) {
+	af := &ArrayField{Type: "array"}
+	return af, af.convertArray(val)
+}
+
+// MarshalJSON implements json marshaller.
+func (af *ArrayField) MarshalJSON() ([]byte, error) {
+	af.Value = make([]interface{}, 0)
+	for _, x := range af.valueFields {
+		af.Value = append(af.Value, x)
+	}
+	for _, x := range af.structFields {
+		af.Value = append(af.Value, x)
+	}
+	for _, x := range af.arrayFields {
+		af.Value = append(af.Value, x)
+	}
+	for _, x := range af.mapFields {
+		af.Value = append(af.Value, x)
+	}
+	type temp ArrayField
+	return json.Marshal((*temp)(af))
+}
+
+func (af *ArrayField) isValid() bool {
+	if af == nil {
+		return false
+	}
+	return len(af.valueFields)+len(af.structFields)+len(af.mapFields)+len(af.arrayFields) > 0
+}
```

Comment:
The "isValid" naming here is a little confusing. Why is an empty array "invalid?" Should the method name here be different? Ditto for the other isValid implementations.

Reply:
hasValue seems better, thanks!